allennlp/common/from_params.py

```diff
@@ -118,6 +125,27 @@ def remove_optional(annotation: type):
         return annotation
 
 
+def infer_params(cls: Type[T], constructor: Callable[..., T] = None):
+    if constructor is None:
+        constructor = cls.__init__
+
+    signature = inspect.signature(constructor)
+    parameters = dict(signature.parameters)
+
+    has_kwargs = False
+    for param in parameters.values():
+        if param.kind == param.VAR_KEYWORD:
+            has_kwargs = True
+
+    if not has_kwargs:
+        return parameters
+
+    super_class = cls.mro()[1]
```

Comment:
Is this correct? Don't we need to check which super class inherits from `Registrable` to be agnostic of inheritance order?

Reply:
I.e I thought this was what @mahnerak had fixed in his PR which this supersedes.

Reply:
@mahnerak was dealing with two problems, which can be seen in the added tests: (1) deeper inheritance trees, where you need to check the superclass's superclass, and (2) when you (for some reason) override the type of an argument in a subclass; in this case the subclass's type annotation should be used for inference, not the superclass's.  That PR did not deal with multiple inheritance, which is what you're worried about here.  That's still a potential problem.

Reply:
Looks like there was a comment that got dropped in that PR, though, which I'm adding back in, and giving a note that this could fail in certain cases of multiple inheritance.

Reply:
would it make sense to just fix the multiple inheritance case here, because we actually do use this frequently with `nn.Module`? 

e.g 

```python
super_class = None
super_classes = cls.mro()
while super_class is None:
	candidate = super_classes.pop(0)

	if issubclass(candidate, Registrable):
		super_class = candidate

```
or something?

Reply:
Hmm, yeah, I see your point.  Let me open a new issue for that.  It should be a pretty easy, self-contained PR to get that to work.